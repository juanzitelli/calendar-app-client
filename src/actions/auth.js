import Swal from "sweetalert2";
import { fetchConToken, fetchSinToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = (email, password) => async dispatch => {
  const respuesta = await fetchSinToken("auth", { email, password }, "POST");
  const body = await respuesta.json();
  if (body.ok) {
    localStorage.setItem("x-token", body.token);
    localStorage.setItem("token-init-date", new Date().getTime());
    dispatch(login({ uid: body.uid, name: body.name }));
  } else {
    Swal.fire("Error", body.msg, "error");
  }
};

export const login = user => ({
  type: types.authLogin,
  payload: user,
});

export const startRegister = (email, password, name) => async dispatch => {
  const respuesta = await fetchSinToken(
    "auth/new",
    { email, password, name },
    "POST"
  );
  const body = await respuesta.json();
  if (body.ok) {
    localStorage.setItem("x-token", body.token);
    localStorage.setItem("token-init-date", new Date().getTime());
    dispatch(login({ uid: body.uid, name: body.name }));
  } else {
    Swal.fire("Error", body.msg, "error");
  }
};

export const startChecking = () => async dispatch => {
  const respuesta = await fetchConToken("auth/renew");
  const body = await respuesta.json();
  if (body.ok) {
    localStorage.setItem("x-token", body.token);
    localStorage.setItem("token-init-date", new Date().getTime());
    dispatch(login({ uid: body.uid, name: body.name }));
  } else {
    dispatch(checkingFinished());
  }
};

const checkingFinished = () => ({
  type: types.authCheckingFinished,
});

export const startLogout = () => async dispatch => {
  localStorage.clear();
  dispatch(logout());
};

export const logout = () => ({
  type: types.authLogout,
});
