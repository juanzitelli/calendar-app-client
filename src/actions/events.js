import Swal from "sweetalert2";
import { fetchConToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

export const eventAddNew = event => ({
  type: types.eventAddNew,
  payload: event,
});

export const startEventAddNew = event => async (dispatch, getState) => {
  const { uid, name } = getState().auth;
  try {
    const resp = await fetchConToken("events", event, "POST");
    const body = await resp.json();

    if (body.ok) {
      event.id = body.evento.id;
      event.user = {
        _id: uid,
        name,
      };
      dispatch(eventAddNew(event));
    }
  } catch (error) {
    console.error(error);
  }
};

export const eventSetActive = event => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent,
});

export const startEventUpdate = event => async dispatch => {
  try {
    const resp = await fetchConToken(`events/${event.id}`, event, "PUT");
    const body = await resp.json();
    if (body.ok) {
      dispatch(eventUpdated(event));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  } catch (error) {
    console.error(error);
  }
};

export const eventUpdated = event => ({
  type: types.eventUpdated,
  payload: event,
});

export const startEventDelete = () => async (dispatch, getState) => {
  const { id } = getState().calendar.activeEvent;
  try {
    const resp = await fetchConToken(`events/${id}`, {}, "DELETE");
    const body = await resp.json();
    if (body.ok) {
      dispatch(eventDeleted());
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  } catch (error) {
    console.error(error);
  }
};

const eventDeleted = () => ({
  type: types.eventDeleted,
});

export const eventsStartLoading = () => async dispatch => {
  try {
    const response = await fetchConToken("events");
    const { eventos } = await response.json();
    
    dispatch(eventsLoaded(prepareEvents(eventos)));
  } catch (error) {
    console.error(error);
  }
};

const eventsLoaded = events => ({
  type: types.eventsLoaded,
  payload: events,
});

export const eventsLogout = () => ({
  type: types.eventLogout,
})
