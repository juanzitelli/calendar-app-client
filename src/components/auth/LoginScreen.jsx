import React from 'react';
import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';
import { startLogin, startRegister } from '../../actions/auth';
import useForm from '../../hooks/useForm';
import './login-screen.css';

export const LoginScreen = () => {

	const dispatch = useDispatch();


	const [formLoginValues, handleLoginInputChange] = useForm({
		loginEmail: '',
		loginPassword: '',
	});

	const { loginEmail, loginPassword } = formLoginValues;

	const handleLogin = (event) => {
		event.preventDefault();
		dispatch(startLogin(loginEmail, loginPassword))
	}



	const [formRegisterValues, handleRegisterInputChange] = useForm({
		registerName: '',
		registerEmail: '',
		registerPassword1: '',
		registerPassword2: '',
	});
	const { registerName,
		registerEmail,
		registerPassword1,
		registerPassword2, } = formRegisterValues


	const handleRegister = (event) => {
		event.preventDefault();
		if (registerPassword1 !== registerPassword2) {
			return Swal.fire('Error', 'Passwords should match', 'error')
		}
		
		dispatch(startRegister(registerEmail, registerPassword1, registerName,))
	}

	return (

		<div className="container login-container">
			<div className="row">
				<div className="col-md-6 login-form-1">
					<h3>Sign In</h3>
					<form onSubmit={handleLogin}>
						<div className="form-group">
							<input
								type="text"
								className="form-control"
								placeholder="Email"
								name="loginEmail"
								value={loginEmail}
								onChange={handleLoginInputChange}

							/>
						</div>
						<div className="form-group">
							<input
								type="password"
								className="form-control"
								placeholder="Password"
								name="loginPassword"
								value={loginPassword}
								onChange={handleLoginInputChange}
							/>
						</div>
						<div className="form-group">
							<input
								type="submit"
								className="btnSubmit"
								value="Login"
							/>
						</div>
					</form>
				</div>

				<div className="col-md-6 login-form-2">
					<h3>Register</h3>
					<form onSubmit={handleRegister}>
						<div className="form-group">
							<input
								type="text"
								className="form-control"
								placeholder="Name"
								name="registerName"
								value={registerName}
								onChange={handleRegisterInputChange}
							/>
						</div>
						<div className="form-group">
							<input
								type="email"
								className="form-control"
								placeholder="Email"
								name="registerEmail"
								value={registerEmail}
								onChange={handleRegisterInputChange}
							/>
						</div>
						<div className="form-group">
							<input
								type="password"
								className="form-control"
								placeholder="Password"	
								name="registerPassword1"
								value={registerPassword1}
								onChange={handleRegisterInputChange}
							/>
						</div>

						<div className="form-group">
							<input
								type="password"
								className="form-control"
								placeholder="Repeat your password"
								name="registerPassword2"
								value={registerPassword2}
								onChange={handleRegisterInputChange}
							/>
						</div>

						<div className="form-group">
							<input
								type="submit"
								className="btnSubmit"
								value="Create account" />
						</div>
					</form>
				</div>
			</div>
		</div>
	)
}