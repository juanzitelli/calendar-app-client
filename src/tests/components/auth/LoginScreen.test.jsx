import "@testing-library/jest-dom";
import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import Swal from 'sweetalert2';
import { startLogin, startRegister } from '../../../actions/auth';
import { LoginScreen } from '../../../components/auth/LoginScreen';

jest.mock('../../../actions/auth', () => ({
	startLogin: jest.fn(),
	startRegister: jest.fn(),
}))
jest.mock('sweetalert2', () => ({
	fire: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn()
const wrapper = mount(
	<Provider store={store}>
		<LoginScreen />
	</Provider>
);



describe('Tests on LoginScreen component', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	})
	test('Should render correctly', () => {
		expect(wrapper).toMatchSnapshot()
	})

	test('Should call login dispatch', () => {
		wrapper.find('input[name="loginEmail"]').simulate("change", {
			target: {
				name: "loginEmail",
				value: "juan@gmail.com"
			}
		})
		wrapper.find('input[name="loginPassword"]').simulate("change", {
			target: {
				name: "loginPassword",
				value: "abc123"
			}
		})

		// wrapper.find('form').at(0).prop('onSubmit', {
		// 	preventDefault() { }
		// })


		wrapper.find("form").at(0).prop("onSubmit")({
			preventDefault() { },
		});

		expect(startLogin).toHaveBeenCalledWith('juan@gmail.com', 'abc123')
	})

	test('Should not startRegister if passwords do not match', () => {
		wrapper.find('input[name="registerPassword1"]').simulate("change", {
			target: {
				name: "registerPassword1",
				value: "123123"
			}
		})
		wrapper.find('input[name="registerPassword2"]').simulate("change", {
			target: {
				name: "registerPassword2",
				value: "123124"
			}
		})

		wrapper.find("form").at(1).prop("onSubmit")({
			preventDefault() { },
		});


		expect(startRegister).not.toHaveBeenCalled();
		expect(Swal.fire).toHaveBeenCalledWith(
			"Error",
			"Passwords should match",
			"error"
		);
	});
	
	test('Should startRegister if passwords match', () => {
		wrapper.find('input[name="registerPassword1"]').simulate("change", {
			target: {
				name: "registerPassword1",
				value: "123123"
			}
		})
		wrapper.find('input[name="registerPassword2"]').simulate("change", {
			target: {
				name: "registerPassword2",
				value: "123123"
			}
		})
		


		wrapper.find("form").at(1).prop("onSubmit")({
			preventDefault() { },
		});



		

		expect(Swal.fire).not.toHaveBeenCalled()
		
		expect(startRegister).toHaveBeenCalledWith(
			"",
			"123123",
			""
		);;
	})
	
})