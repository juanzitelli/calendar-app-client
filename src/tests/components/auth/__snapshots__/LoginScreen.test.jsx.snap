// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on LoginScreen component Should render correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Sign In
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Email"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword1"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword2"
            onChange={[Function]}
            placeholder="Repeat your password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Create account"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
